pkg/labels/selector.go

```diff
@@ -76,109 +76,35 @@ func NewSelector() Selector {
 	return internalSelector(nil)
 }
 
-type internalSelector []Requirement
-
-// Sort by key to obtain determisitic parser
-type ByKey []Requirement
-
-func (a ByKey) Len() int { return len(a) }
-
-func (a ByKey) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
-
-func (a ByKey) Less(i, j int) bool { return a[i].key < a[j].key }
-
-// Requirement is a selector that contains values, a key
-// and an operator that relates the key and values. The zero
-// value of Requirement is invalid.
-// Requirement implements both set based match and exact match
-// Requirement is initialized via NewRequirement constructor for creating a valid Requirement.
-type Requirement struct {
-	key       string
-	operator  Operator
-	strValues sets.String
-}
+type internalSelector []selectors.Requirement
 
 // NewRequirement is the constructor for a Requirement.
+// It provides additional validation over selectors.NewRequirement as follows:
 // If any of these rules is violated, an error is returned:
-// (1) The operator can only be In, NotIn, Equals, DoubleEquals, NotEquals, Exists, or DoesNotExist.
-// (2) If the operator is In or NotIn, the values set must be non-empty.
-// (3) If the operator is Equals, DoubleEquals, or NotEquals, the values set must contain one value.
-// (4) If the operator is Exists or DoesNotExist, the value set must be empty.
-// (5) The key is invalid due to its length, or sequence
+// (1) The key is invalid due to its length, or sequence
 //     of characters. See validateLabelKey for more details.
+// (2) The value is invalid, See validateLabelValue for more details.
 //
 // The empty string is a valid value in the input values set.
-func NewRequirement(key string, op Operator, vals sets.String) (*Requirement, error) {
+func NewRequirement(key string, op selectors.Operator, vals sets.String) (*selectors.Requirement, error) {
 	if err := validateLabelKey(key); err != nil {
 		return nil, err
 	}
-	switch op {
-	case InOperator, NotInOperator:
-		if len(vals) == 0 {
-			return nil, fmt.Errorf("for 'in', 'notin' operators, values set can't be empty")
-		}
-	case EqualsOperator, DoubleEqualsOperator, NotEqualsOperator:
-		if len(vals) != 1 {
-			return nil, fmt.Errorf("exact-match compatibility requires one single value")
-		}
-	case ExistsOperator, DoesNotExistOperator:
-		if len(vals) != 0 {
-			return nil, fmt.Errorf("values set must be empty for exists and does not exist")
-		}
-	default:
```

Comment:
Please note that there are two new operators ('Gt' and 'Lt') in `pkg/label/selector.go`, you may need manually move the corresponding logic to the new place when rebasing to latest master.